### ISSUE TITLE
Add support for Swagger (OpenApi) 3.0

### ISSUE THREAD

**Issue by cdimascio** [APIDevTools/swagger-express-middleware#81]
This is an excellent project, it will be great to see support for the new spec version.

https://github.com/OAI/OpenAPI-Specification/blob/3.0.0-rc0/versions/3.0.md
https://github.com/OAI/OpenAPI-Specification

**Comment by BigstickCarpet** [comment]
I fully intend to add support for OAI 3.0 soon 👍

**Comment by tyrsius** [comment]
@BigstickCarpet any update on this?

**Comment by BigstickCarpet** [comment]
Yes, I'm making good progress on this.  It's a lot of work and involves [multiple](https://github.com/BigstickCarpet/swagger-parser) [different](https://github.com/BigstickCarpet/json-schema-ref-parser/) [projects](https://github.com/BigstickCarpet/json-schema-lib), so it will take some time.

**Comment by Stwissel** [comment]
Where can we help out?

**Comment by BigstickCarpet** [comment]
You can review my progress [here](https://github.com/BigstickCarpet/json-schema-lib) and give me any feedback that you have.  I'm currently on a month-long business trip, so I haven't been able to work on it lately, but I'll get right back to it as soon as I return home in late September.

**Comment by kevingilbert100** [comment]
Excited for updates, will contribute if I can.

**Comment by vtremblay** [comment]
@BigstickCarpet are things going well for this update? Is there anything we can do?

**Comment by danlevan** [comment]
We would like this as well :)

**Comment by BigstickCarpet** [comment]
Hi all.  I'm sorry for the delay.  I had to put this project on hold for a couple months due to other commitments.  So I basically made no progress in September or October.  But now I'm back on the job and really hoping to get something released ASAP.  I'm aware that there are a *lot* of people waiting on me, including some very big projects.  I feel the pressure.  And I promise I'm moving as fast as I possibly can. 😥

**Comment by gonenduk** [comment]
Any updates?

**Comment by gonenduk** [comment]
bump

**Comment by gonenduk** [comment]
bump

**Comment by gonenduk** [comment]
Hey @BigstickCarpet,
Now that swagger-parser 5.0 supports OpenAPI 3, can we have an update for this package as well?

**Comment by BigstickCarpet** [comment]
Yes, I plan to upgrade this project next.  No promises on _when_ though.  I'm completely swamped at work right now, so I don't have a lot of spare time to work on this.  I'll get it done as soon as I possibly can though.

**Comment by gonenduk** [comment]
Is there any way we can help?

**Comment by ronkorving** [comment]
Hi :) Is this still going places?

**Comment by jefflage** [comment]
I'm sure there is a great deal more to it, but locally I'm running a form where for all of the param parsing, I've replaced function calls that use `param` (where param is the JSON schema object of the parameter) with `param.schema || param` and things seem to be working really well.  I went with this approach because it seemed the easiest way to backwards compatible.

**Comment by gonenduk** [comment]
@jefflage - can you share your work with us?
Did you fork and made some changes?

**Comment by jefflage** [comment]
Not yet.  Just hacked at the source locally.  But now that we’re further into the spike we intend to fork and submit a PR.  I’ll post when we do.

**Comment by gonenduk** [comment]
@jefflage - can't wait! And probably also the other 28K people who download it every week!

**Comment by ronkorving** [comment]
@gonenduk "no pressure" :D

**Comment by gonenduk** [comment]
No pressure at all, but really can't wait to test it and get rid of version 2.

**Comment by gonenduk** [comment]
Version 1.0.0 is out! Still not support for OpenAPI 3, but it seems the project is getting an attention now and many things are being fixed.

**Comment by jdpowell1** [comment]
Any updates on this? Would be great to use on an OpenAPI 3.0 documented API that I'm work on!

**Comment by eszjay** [comment]
@jdpowell1 Markus has made some good progress in his fork https://github.com/APIDevTools/swagger-express-middleware/pull/119 which he is publishing to NPM https://www.npmjs.com/package/swagger-express-middleware3 . There is also another project with a bit of a different feature set that may or may not be helpful for your project: https://github.com/exegesis-js/exegesis .

**Comment by vtremblay** [comment]
Anything new on this?

**Comment by JamesMessinger** [comment]
I've been busy recently and unable to work on this, but I did notice that @sedgewickmm18 submitted PR #119 which adds OpenAPI 3.0 support.  As soon as I get a chance, I'll review that PR and possibly merge it.

**Comment by gonenduk** [comment]
Hey.
Any updates? Maybe news about PR #119?

**Comment by sedgewickmm18** [comment]
@gonenduk I've rebased PR #119 but I'm unsure how to move on. This is still only the beginning, as additional test cases are needed to cover API 3.0 specifics.

**Comment by JamesMessinger** [comment]
I just now merged the PR.   As @sedgewickmm18 said, it's only the beginning and additional OAS3 stuff will need to be added, but this is a great start!

**Comment by gonenduk** [comment]
Thanks for the effort guys!

I tried to use the merge and found a problem with basePath. With OpenAPI basePath is no longer used. Instead, a list of possible servers are used.
I created a PR #136 to fix this.
Please review it @JamesMessinger, @sedgewickmm18

**Comment by JamesMessinger** [comment]
We've created [a new branch](https://github.com/APIDevTools/swagger-express-middleware/tree/openapi-3-support) to work on OpenAPI 3.0 support.

**Comment by gonenduk** [comment]
Great news!
Can you publish that branch to npm as a beta?
It will allow me and others to know about it, test it and report back any issues we find.

**Comment by airen29** [comment]
looking forward for the update :)

**Comment by mdaehn** [comment]
Nice work on this NPM package. Our team is planning to use it, but with only one little snag, the OpenAPI 3 support. Wondering if it is safe to pull down your branch and build it locally, so we can start using it. Do you have an estimate when this will be complete?

**Comment by JamesMessinger** [comment]
OpenAPI 3 support is important to me too, since I have a bunch of Swagger 2.0 APIs that I would *love* to upgrade to OpenAPI 3.  I can't give an ETA though, because it totally depends on how much free time I have to work on it.  This is a nights-and-weekends project for me.

For the time being, a workaround would be to define your APIs using Swagger 2.0, which is fully supported.  Then once I get OpenAPI 3 support done, you can easily convert your Swagger 2.0 definitions to OpenAPI 3 using a library such as [swagger2openapi](https://www.npmjs.com/package/swagger2openapi) or the online [API Transformer tool](https://www.apimatic.io/transformer).

**Comment by mdaehn** [comment]
Thank James. Since I am in the starting phase, I just forked your `openapi-3-support` branch and added the 3.0 validation pertinent to my project. I just run my code with the forked version until your are able to make the updates.

**Comment by JamesMessinger** [comment]
We have started publishing **alpha** releases to npm with some support for OpenAPI 3.0.  These alphas are based on [the v3 branch](https://github.com/APIDevTools/swagger-express-middleware/tree/v3).  See [the ReadMe](https://github.com/APIDevTools/swagger-express-middleware/tree/v3#readme) and [the changelog](https://github.com/APIDevTools/swagger-express-middleware/blob/v3/CHANGELOG.md) for more details.

**Comment by kazazor** [comment]
Hi @JamesMessinger , great job so far on OAS 3 :)

I was trying to look at the git issues to evaluate how much work got left in order to support it but it's not marked with any label. I did find the label `OpenAPI 3` but it was on only 1 github issue while there are more + I did not see a backlog in any of the issues but only bugs/fixes.

Can you share your perspective on the amount of work/tasks that still exists on OAS 3?
I like the package and would love it to be the first truly OAS 3 supported one.

**Comment by gorjuspixels** [comment]
@kazazor it looks like validate middleware is not working yet (a few are missing `next()` calls and the one that validates request body is not yet implemented). Not sure if there's more

**Comment by gonenduk** [comment]
Any update on this 2.5 years issue?

**Comment by JamesMessinger** [comment]
Update here: https://github.com/APIDevTools/swagger-express-middleware/issues/158
